### PR TITLE
New platform: Intel Galileo Board

### PR DESCRIPTION
**Pull request by aguedes**
Hi all,

This PR is a follow-up from the Galileo port effort announced a couple of weeks ago in Contiki's mailing list (for further information check this [link](http://sourceforge.net/p/contiki/mailman/message/34350483/)).

So this PR adds support for both Intel Quark X1000 SoC (cpu layer) and the Intel Galileo Board (platform layer). In order to assist the review process, below follows some comments that summarize the patches within this PR.

* Cpu layer patches are prefixed with "x86:" and implement the following features:
  * IA initialization (GDT, IDT and IRQs)
  * Programmable Interrupt Controller (PIC) driver
  * Real-time Clock (RTC) driver
  * Programmable Interval Timer (PIT) driver
  * UART driver
  * PCI driver
* Platform layer patches are prefixed with "galileo:" and implement the following features:
  * Clock module
  * Etimer, Ctimer and Rtimer libraries
  * Stdio library (stdout and stderr)
* The patch "examples: Add all-timers example" introduces a Contiki sample application to validate the timers implementation. This sample is not Galileo-specific and it can be used by any port.
* BSP-related files are located in platform/galileo/bsp/ directory. We provide shell scripts that downloads and build the BSP components.
* platform/galileo/README.md provides detailed instructions on how to build, run and debug application for the platform.

This PR doesn't provide Travis integration. Apart from that, it addresses all others requirements described in the [New Platforms section](https://github.com/contiki-os/contiki/wiki/Code-Contributions#new-platforms) from the contribution guide.

Looking forward to get your feedback!

Best regards,

Andre

### CODE REVIEW

**Comment by jeez** [comment]
Since Github re-organizes the commits in chronological order, reviewing this PR on its original branch might be easier. It starts at --> https://github.com/otcshare/contiki-x86/commits/x86?page=2 .

**Comment by alignan** [comment]
I have a Galileo at home, I will test it anytime next week.

**Comment by jeez** [comment]
[**@alignan**](https://github.com/alignan) commented on [Aug 20, 2015, 12:57 PM GMT-3](https://github.com/contiki-os/contiki/pull/1196#issuecomment-133059552 "2015-08-20T15:57:35Z - Replied by Github Comment Enhancer"):
> I have a Galileo at home, I will test it anytime next week.

Great, thanks! The README file (https://github.com/otcshare/contiki-x86/blob/x86/platform/galileo/README.md) should provide everything you need. Please let us know in case there is something missing.

**Comment by jeez** [comment]
Hi @alignan , did you have any chance to test the port already?

**Comment by alignan** [comment]
Nope, not yet but planning soon.  As this requires some of my free time I'm not rushing into this, but I definitely want to give it a go :-)

**Comment by jeez** [comment]
We have some more stuff coming for this port, so it would be great if more people could have a look at it. :) For instance, we already have UEFI support [1] and an Ethernet driver [2], and there are other drivers being cooked as we speak!

[1] https://github.com/otcshare/contiki-x86/pull/2
[2] https://github.com/otcshare/contiki-x86/pull/1

File: .gitignore [hunk 1]
```diff
@@ -92,3 +92,21 @@ contiki-cc2530dk.lib
 # rl78 build artifacts
 *.eval-adf7xxxmb4z
 *.eval-adf7xxxmb4z.srec
+
+# cscope files
+cscope.*
+
+# vim swap files
```
Comment (alignan) [inline .gitignore]: Only include PR-specifics

**Comment by alignan** [comment]
Bulding with `./platform/galileo/bsp/libc/build_newlib.sh` gives an error:

`cp: cannot stat ‘./newlib-2.2.0-1/install/i586-elf’: No such file or directory`

In case you want to review the build instructions and list missing dependencies (like `binutils`), etc.

File: .gitignore (hunk 1 above)
Comment (aguedes) [inline .gitignore]: Sure, I'll remove the gitignore patch from this PR.

I'll send it in a separated PR since it might be useful for others.

**Comment by aguedes** [comment]
Sure, I'll revise the list of dependencies and update it according. Thanks!

Did you manage to get newlib built?

**Comment by alignan** [comment]
I didn't test further, as this is a problem a new user might found upon testing the Galileo, I prefer to have the build process updated accordingly or a suggestion in the `README` file if this is a common error.  Mostly to avoid the same questions on the mailing list.

**Comment by jeez** [comment]
[**@alignan**](https://github.com/alignan) commented on [Sep 26, 2015, 3:13 PM GMT-3](https://github.com/contiki-os/contiki/pull/1196#issuecomment-143477180 "2015-09-26T18:13:59Z - Replied by Github Comment Enhancer"):
> Bulding with `./platform/galileo/bsp/libc/build_newlib.sh` gives an error:
> 
> `cp: cannot stat ‘./newlib-2.2.0-1/install/i586-elf’: No such file or directory`

Sorry, I'm not sure I followed. Just tested this command from contiki root folder and it works.
Are you saying it fails if ran from platform/galileo/ ?


> 
> In case you want to review the build instructions and list missing dependencies (like `binutils`), etc.

Dependencies are always tricky to list since they most likely change from one distro/env to another.
We must provide a high level overview of what is currently needed, yes, but we will still get questions every now and then related to that.

**Comment by alignan** [comment]
I ran the command from contiki root folder too but failed as described before.  I will try to compile `newlib` again with a different set of instructions to see what happens.  Nevertheless, what distro did you use for testing?

**Comment by jeez** [comment]
[**@alignan**](https://github.com/alignan) commented on [Sep 30, 2015, 4:08 AM GMT-3](https://github.com/contiki-os/contiki/pull/1196#issuecomment-144310216 "2015-09-30T07:08:41Z - Replied by Github Comment Enhancer"):
> I ran the command from contiki root folder too but failed as described before.  I will try to compile `newlib` again with a different set of instructions to see what happens.  Nevertheless, what distro did you use for testing?

I use ArchLinux. What is your distro?

**Comment by alignan** [comment]
Ubuntu 32 and 64 bits.

**Comment by joakimeriksson** [comment]
I also tried building and managed to do the first step in Instant Contiki 3.0 but then other steps (UEFI) did only support 64-bits environments so that failed. And when trying on a 64-bit ubuntu I got the same error as @alignan . Is there any way you could get this working in Instant Contiki 3.0? That would be the best place to get it working since next Instant Contiki will anyway need to build it - and they are usually a 32-bit Ubuntu (at least so far).

**Comment by jeez** [comment]
@mlemay-intel can you please have a look at the issue @joakimeriksson is describing above?

**Comment by joakimeriksson** [comment]
With the support for old versions of Make it is now possible to build it in Instant Contiki 3.0. @alignan can you also make a quick re-try?

**Comment by alignan** [comment]
Sure! I will test this tomorrow when I have the hardware available

**Comment by alignan** [comment]
No luck unfortunately, tested on an Ubuntu 64 bit machine, fails at `make -j5 all`

**Comment by aguedes** [comment]
Could you please copy the output to pastebin or something so we can take a look?

**Comment by joakimeriksson** [comment]
@alignan Did you also try in Instan Contiki 3.0?

**Comment by alignan** [comment]
Sorry, I meant to update my previous post.
I compiled `newlib` ok but I needed to install also `texinfo` support.
Then upon building `UEFI` I stumbled upon `cannot find -luuid`, fixable by installing the `uuid-dev` package... then:
```
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/4.8/libgcc.a when searching for -lgcc
/usr/bin/ld: cannot find -lgcc
```
Fixed by:
```
sudo apt-get install gcc-multilib g++-multilib
```
Then joy:
```
  AR        contiki-galileo.a
  CC        hello-world.c
  CC        ../../platform/galileo/./newlib-syscalls.c
  LD        hello-world.galileo
  LD        hello-world.galileo.dll
   text	   data	    bss	    dec	    hex	filename
  80315	   1584	  21000	 102899	  191f3	hello-world.galileo
  85824	  22688	      0	 108512	  1a7e0	hello-world.galileo.efi
rm hello-world.galileo.dll obj_galileo/newlib-syscalls.o hello-world.co
```
I'm missing only running on the actual hardware, I will try to do this over the weekend (I hope to find time), but is a step forward :-)
Please include the above steps in the `README` file as well.

**Comment by alignan** [comment]
Out of curiosity, how far is this port to support the Edison as well?

**Comment by aguedes** [comment]
Thanks @alignan! I'll update the README file with your input.

I'm not very familiar with Edison specs so, to be honest, I'm not sure how far it is.

**Comment by jeez** [comment]
Edison is a very different platform. I would say this port is as far from it as how you would want to run it there.

If it's for running on the Atom core, then we would need to be able to boot from u-boot, since Edison does not provide UEFI firmware as far as I know.

If it's for running on the MCU there, then, well, we are super far. :)

**Comment by aguedes** [comment]
Hi @alignan, the prerequisite packages you reported were already included in the README (thanks @mlemay-intel!).

**Comment by alignan** [comment]
I'm done with testing, I don't have the UART jack cable but overall it looks OK.  As a general comment, the link `[2]` in the `README` is redirected to a standard support page, but browsing the cached version shows the proper page.

So the last step for the merging to be done is to ran uncrustify and fix coded style issues (no space in `if`, etc), this is the list of files you should fix:

````
platform/galileo/contiki-main.c
platform/galileo/drivers/galileo-pinmux.c
platform/galileo/drivers/gpio-pcal9535a.c
platform/galileo/drivers/pwm-pca9685.c
````
You can use `uncrustify` from the `tools/code-style` location.

And that's it.

**Comment by mlemay-intel** [comment]
Thank you for testing this, @alignan.
I prepared a patch to correct the broken link: 4d81c20276f34a40cab73be159bca97cdf1186b9  I also prepared a patch to fix some of the code style violations: aae84366eb8cad7b90a42818b92beee2f0590705  I plan to apply these patches to this PR after reviewing them with my colleagues at Intel, unless you would like additional changes.  In particular, I omitted some of the code changes suggested by uncrustify, since I thought they would actually make the code less readable.  However, I will add those changes if you think I should.  Here are the suggestions from uncrustify that I disagreed with:
```
--- platform/galileo/contiki-main.c	2015-12-19 17:23:24.127190851 -0800
+++ /dev/stdin	2015-12-19 17:30:46.760711006 -0800
@@ -37,8 +37,8 @@
 #include "uart.h"
 #include "eth-conf.h"
 
-PROCINIT(  &etimer_process
-         , &tcpip_process
+PROCINIT(&etimer_process,
+         &tcpip_process
 #if WITH_DNS
          , &resolv_process
 #endif
```
The change above would disrupt the alignment of the process variable names.
```
--- platform/galileo/drivers/galileo-pinmux.c	2015-12-19 17:26:35.592047664 -0800
+++ /dev/stdin	2015-12-19 17:31:19.025868101 -0800
@@ -63,16 +63,16 @@
 };
 
 static struct pin_config default_pinmux_config[PINMUX_NUM_PINS] = {
-  { 0,  GALILEO_PINMUX_FUNC_C }, /* UART0_RXD */
-  { 1,  GALILEO_PINMUX_FUNC_C }, /* UART0_TXD */
-  { 2,  GALILEO_PINMUX_FUNC_A }, /* GPIO5(out) */
...
+  { 0, GALILEO_PINMUX_FUNC_C },  /* UART0_RXD */
+  { 1, GALILEO_PINMUX_FUNC_C },  /* UART0_TXD */
+  { 2, GALILEO_PINMUX_FUNC_A },  /* GPIO5(out) */
...
   { 10, GALILEO_PINMUX_FUNC_A }, /* GPIO2 (out) */
   { 11, GALILEO_PINMUX_FUNC_B }, /* GPIO_SUS3 (in) */
   { 12, GALILEO_PINMUX_FUNC_B }, /* GPIO7 (in) */
@@ -108,425 +108,505 @@
 static struct pinmux_internal_data data;
 
 static struct mux_path galileo_pinmux_paths[PINMUX_NUM_PINS * PINMUX_NUM_FUNCS] = {
-  {0, GALILEO_PINMUX_FUNC_A, {
-          { EXP1,  0,  PIN_HIGH, (QUARKX1000_GPIO_OUT) }, /* GPIO3 out */
-          { EXP1,  1,   PIN_LOW, (QUARKX1000_GPIO_OUT) },
-          { NONE,  0,  DISABLED, (QUARKX1000_GPIO_IN ) },
-          { NONE,  0,  DISABLED, (QUARKX1000_GPIO_IN ) }}},
...
-  {19, GALILEO_PINMUX_FUNC_D, {
-           { NONE,  0,  DISABLED, (QUARKX1000_GPIO_IN ) }, /* NONE */
-           { NONE,  0,  DISABLED, (QUARKX1000_GPIO_IN ) },
-           { NONE,  0,  DISABLED, (QUARKX1000_GPIO_IN ) },
-           { NONE,  0,  DISABLED, (QUARKX1000_GPIO_IN ) }}},
+  { 0, GALILEO_PINMUX_FUNC_A, {
+      { EXP1, 0, PIN_HIGH, (QUARKX1000_GPIO_OUT) },       /* GPIO3 out */
+      { EXP1, 1, PIN_LOW, (QUARKX1000_GPIO_OUT) },
+      { NONE, 0, DISABLED, (QUARKX1000_GPIO_IN) },
+      { NONE, 0, DISABLED, (QUARKX1000_GPIO_IN) }
+    } },
...
+  { 19, GALILEO_PINMUX_FUNC_D, {
+      { NONE, 0, DISABLED, (QUARKX1000_GPIO_IN) },         /* NONE */
+      { NONE, 0, DISABLED, (QUARKX1000_GPIO_IN) },
+      { NONE, 0, DISABLED, (QUARKX1000_GPIO_IN) },
+      { NONE, 0, DISABLED, (QUARKX1000_GPIO_IN) }
+    } },
 };
 
 int
```
The change above is very long, so I shortened it with ellipses.  This change would disrupt the alignment of the structure initialization expressions.
Please let me know if you would like any additional changes beyond those in my existing patches.

**Comment by alignan** [comment]
I'm OK with the changes in the patch, as you also stated I'm OK to bend a bit the code style to make the code friendly to read :smile:

**Comment by jeez** [comment]
@alignan, thanks for all the testing and reviews!

@mlemay-intel, I didn't find a PR for those commits on our fork, so I gave my LGTM on them directly. Please feel free to push them to the branch so the PR gets updated.

Let's land this, folks! :)

**Comment by jeez** [comment]
@mlemay-intel and thanks for the prompt fixes, by the way!

**Comment by mlemay-intel** [comment]
I updated this PR so it should be ready to merge! :)

**Comment by alignan** [comment]
Great! :+1: from me! @simonduq this can be merged as it is, or rebasing is needed due the amount of commits?

**Comment by simonduq** [comment]
Can me merged as it is AFAIC, but if the authors are willing to rebase/squash first that is very welcome of course!

**Comment by jeez** [comment]
We'd rather keep the commit history as is, if that is OK for you all,
folks. It tells a nice story, IMO. ;)

It seems to be a clean merge the way it is, but please let us know if
otherwise!
Em 20/12/2015 19:06, "Simon Duquennoy" <notifications@github.com> escreveu:

**Comment by jeez** [comment]
@alignan @simonduq @mlemay-intel    I rebased it just now.

All ready to go!

**Comment by alignan** [comment]
Where did you rebased?

**Comment by jeez** [comment]
@alignan what do you mean?
This PR is from branch https://github.com/otcshare/contiki-x86/tree/x86 , which I rebased on top of the master branch from contiki commit hash 8218b70a0d07d53b110f9e25643f597dd0033515 .

**Comment by jeez** [comment]
Just noticed it updated all commits here on the PR, not sure why, though :/ .

**Comment by alignan** [comment]
Ping me when it's done, is all green now :smiley:

**Comment by jeez** [comment]
@alignan it is done!

**Comment by alignan** [comment]
I meant rebasing/squash as @simonduq suggested, sorry for the confusion! (holidays + flu), I thought you were squashing also.

**Pull request merged**